### PR TITLE
Add support for non-decode CUDA graphs for Mamba models

### PR DESCRIPTION
**Pull request by santhnm2**
Adds support for non-decode CUDA graphs for Mamba models.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/contexts/attention_context/triton/tensor_ops.py [hunk 1]
```diff
@@ -0,0 +1,446 @@
+from typing import Optional
+
+import torch
```
Comment (kvareddy) [inline megatron/core/inference/contexts/attention_context/triton/tensor_ops.py:5]: What is the source of these kernels?

File: megatron/core/inference/batch_dimensions_utils.py [hunk 2]
```diff
@@ -145,25 +167,46 @@ def adjust_batch_dims_for_expert_parallelism(
...
+        any_ep_rank_has_explicit_chunked_prefill_req = sync_tensor[2].item() == 1
+
+        # We force eager mode for scenarios where some ranks will run with CUDA graphs
+        # while others will not. Without this check, the all-to-all communication in the
+        # expert routing layer would pad up to the maximum capacity only for the ranks that
+        # are using CUDA graphs in this step, leading to a NCCL hang.
+        # This can happen in the following cases:
+        #   1. If we only allow decode CUDA graphs but some ranks are running non-decode batches
+        #   2. Some ranks are running explicit chunked prefill requests
+        #       (graphs are not recorded for batches with explicit chunked prefill requests)
```
Comment (santhnm2) [inline megatron/core/inference/batch_dimensions_utils.py:197]: @sidsingh-nvidia Is this actually true that the all-to-all will hang if there's a mismatch in which ranks are running with CUDA graphs? If there's a dummy forward pass won't that always have to run without CUDA graphs because we don't pass an `inference_context`? Which would suggest it's ok for ranks to not all be running with CUDA graphs?

Comment (santhnm2) [reply megatron/core/inference/batch_dimensions_utils.py:197]: This might also have implications for whether we need to do this explicit chunked prefill sync

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 3]
```diff
@@ -533,7 +536,9 @@ def _dynamic_step_context_init(
                 )
 
         # Get flat tokens, position ids.
-        if construct_graph_dimensions is not None:
+        # If we are running a dummy forward step we want to use the token count agreed upon
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py:539]: Ah, good catch!

File: megatron/core/inference/batch_dimensions_utils.py (hunk 2 above)
Comment (sidsingh-nvidia) [reply megatron/core/inference/batch_dimensions_utils.py:197]: Not sure what you mean here. I am explicitly running dummy cuda graphed forwards, which do have reference to the inference context. Refer to the logic in text controller.

File: megatron/core/inference/batch_dimensions_utils.py [hunk 4]
```diff
@@ -279,6 +321,8 @@ def generate_cuda_graph_batch_dimensions_list(
             max_tokens: Maximum total tokens
             max_sequence_length: Maximum sequence length
             use_cuda_graphs_for_non_decode_steps: Whether to use CUDA graphs for non-decode steps
+            requires_explicit_chunked_prefill_req: Whether the model requires an explicit chunked
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/batch_dimensions_utils.py]: Is this something you forgot to remove?

File: megatron/core/inference/batch_dimensions_utils.py (hunk 2 above)
Comment (santhnm2) [reply megatron/core/inference/batch_dimensions_utils.py:197]: Discussed offline, will resolve

File: megatron/core/inference/batch_dimensions_utils.py (hunk 4 above)
Comment (santhnm2) [reply megatron/core/inference/batch_dimensions_utils.py]: Yes, thanks for catching!
